### PR TITLE
[cloud-provider-azure] Virtual Network service endpoints

### PR DESCRIPTION
**Pull request by vitaliy-sn**
## Description

Added `serviceEndpoints` parameter in provider cluster configuration for Azure cloud provider.

## Why do we need it, and what problem does it solve?

This will allow you to connect external azure services to the cluster network.

## Manual test results
<details><summary>`serviceEndpoints: [Microsoft.Storage]` has been added to providerClusterConfiguration</summary>

```
│ │ │   # azurerm_subnet.kube will be updated in-place
│ │ │   ~ resource "azurerm_subnet" "kube" {
│ │ │         address_prefix                                 = "10.50.0.0/24"
│ │ │         address_prefixes                               = [
│ │ │             "10.50.0.0/24",
│ │ │         ]
│ │ │         enforce_private_link_endpoint_network_policies = false
│ │ │         enforce_private_link_service_network_policies  = false
│ │ │         id                                             = "/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroy ↵
│ │ │ meafter30aug/providers/Microsoft.Network/virtualNetworks/destroymeafter30aug/subnets/destroymeafter30aug"
│ │ │         name                                           = "destroymeafter30aug"
│ │ │         resource_group_name                            = "destroymeafter30aug"
│ │ │       ~ service_endpoints                              = [
│ │ │           + "Microsoft.Storage",
│ │ │         ]
│ │ │         virtual_network_name                           = "destroymeafter30aug"
│ │ │     }
│ │ │ 
│ │ │ Plan: 0 to add, 1 to change, 0 to destroy.
```

image

</details>

<details><summary>The bootstrap is made from the main branch. `dhctl terraform check` was run from the docker image from this PR.</summary>

```
[deckhouse] root@e1267ddd825a / # dhctl terraform check --ssh-agent-private-keys /tmp/.ssh/id_rsa_tfadm --ssh-user azureuser --ssh-host 20.56.4.235
Check started ...

Enter passphrase for /tmp/.ssh/id_rsa_tfadm: 
ssh-add: Identity added: /tmp/.ssh/id_rsa_tfadm (/tmp/.ssh/id_rsa_tfadm)
Certificate added: /tmp/.ssh/id_rsa_tfadm-cert.pub                                                                                         ↵
(vault-oidc-8d0dcfc1-b131-4c8e-987e-3ef8a2a639e9-268c9a598933feed92182f0710dc4ee7ff409d6e9dfc17af0da719b0ff75c069)

┌ 🎈 ~ Common: Connect to Kubernetes API
│ ┌ Waiting for SSH connection
│ │ Try to connect to 20.56.4.235 host
│ │ 🎉 Succeeded!
│ └ Waiting for SSH connection (1.47 seconds)
│ 
│ ┌ Get Kubernetes API client
│ │ ┌ Starting kube proxy
│ │ │ Using host 20.56.4.235
│ │ │ Got proxy port = 36503 on host 20.56.4.235
│ │ │ 🎉 Succeeded!
│ │ └ Starting kube proxy (2.78 seconds)
│ │ 
│ │ Proxy started on port 22322
│ │ 🎉 Succeeded!
│ └ Get Kubernetes API client (2.79 seconds)
│ 
│ ┌ Waiting for Kubernetes API to become Ready
│ │ 🎉 Succeeded!
│ └ Waiting for Kubernetes API to become Ready (0.16 seconds)
└ 🎈 ~ Common: Connect to Kubernetes API (5.46 seconds)

┌ Get Cluster UUID from the Kubernetes cluster
│ 🎉 Succeeded!
└ Get Cluster UUID from the Kubernetes cluster (0.08 seconds)

┌ Get Cluster Terraform state from Kubernetes cluster
│ 🎉 Succeeded!
└ Get Cluster Terraform state from Kubernetes cluster (0.08 seconds)

┌ 🌱 ~ Terraform: Check state base-infrastructure for Kubernetes cluster
│ ┌ terraform init ...
│ │ 
│ │ Initializing the backend...
│ │ 
│ │ Initializing provider plugins...
│ │ - Finding latest version of hashicorp/azurerm...
│ │ - Installing hashicorp/azurerm v2.36.0...
│ │ - Installed hashicorp/azurerm v2.36.0 (unauthenticated)
│ │ 
│ │ The following providers do not have any version constraints in configuration,
│ │ so the latest version was installed.
│ │ 
│ │ To prevent automatic upgrades to new major versions that may contain breaking
│ │ changes, we recommend adding version constraints in a required_providers block
│ │ in your configuration, with the constraint strings suggested below.
│ │ 
│ │ * hashicorp/azurerm: version = "~> 2.36.0"
│ │ 
│ │ Terraform has been successfully initialized!
│ │ Terraform runner "base-infrastructure" process exited.
│ └ terraform init ... (0.38 seconds)
│ 
│ ┌ terraform plan ...
│ │ Refreshing Terraform state in-memory prior to plan...
│ │ The refreshed state will be used to calculate this plan, but will not be
│ │ persisted to local or remote state storage.
│ │ 
│ │ azurerm_resource_group.kube: Refreshing state...                                                                                       ↵
│ │ [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter31aug]
│ │ azurerm_route_table.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter31a ↵
│ │ ug/providers/Microsoft.Network/routeTables/destroymeafter31aug]
│ │ azurerm_virtual_network.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafte ↵
│ │ r31aug/providers/Microsoft.Network/virtualNetworks/destroymeafter31aug]
│ │ azurerm_network_security_group.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destro ↵
│ │ ymeafter31aug/providers/Microsoft.Network/networkSecurityGroups/destroymeafter31aug]
│ │ azurerm_subnet.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter31aug/pr ↵
│ │ oviders/Microsoft.Network/virtualNetworks/destroymeafter31aug/subnets/destroymeafter31aug]
│ │ azurerm_network_security_rule.icmp: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroy ↵
│ │ meafter31aug/providers/Microsoft.Network/networkSecurityGroups/destroymeafter31aug/securityRules/AllowIcmp]
│ │ azurerm_network_security_rule.ssh: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroym ↵
│ │ eafter31aug/providers/Microsoft.Network/networkSecurityGroups/destroymeafter31aug/securityRules/AllowSsh]
│ │ azurerm_subnet_route_table_association.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroup ↵
│ │ s/destroymeafter31aug/providers/Microsoft.Network/virtualNetworks/destroymeafter31aug/subnets/destroymeafter31aug]
│ │ azurerm_subnet_network_security_group_association.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/re ↵
│ │ sourceGroups/destroymeafter31aug/providers/Microsoft.Network/virtualNetworks/destroymeafter31aug/subnets/destroymeafter31aug]
│ │ 
│ │ ------------------------------------------------------------------------
│ │ 
│ │ No changes. Infrastructure is up-to-date.
│ │ 
│ │ This means that Terraform did not detect any differences between your
│ │ configuration and real physical resources that exist. As a result, no
│ │ actions need to be performed.
│ │ Terraform runner "base-infrastructure" process exited.
│ └ terraform plan ... (13.15 seconds)
└ 🌱 ~ Terraform: Check state base-infrastructure for Kubernetes cluster (14.76 seconds)

┌ Get Nodes Terraform state from Kubernetes cluster
│ nodeGroup=master nodeName=destroymeafter31aug-master-0 symbols=13200
│ 🎉 Succeeded!
└ Get Nodes Terraform state from Kubernetes cluster (0.08 seconds)

┌ Get NodeGroups node template settings
│ 🎉 Succeeded!
└ Get NodeGroups node template settings (0.08 seconds)

┌ 🌱 ~ Terraform: Check state master-node for destroymeafter31aug-master-0
│ ┌ terraform init ...
│ │ 
│ │ Initializing the backend...
│ │ 
│ │ Initializing provider plugins...
│ │ - Using previously-installed hashicorp/azurerm v2.36.0
│ │ 
│ │ The following providers do not have any version constraints in configuration,
│ │ so the latest version was installed.
│ │ 
│ │ To prevent automatic upgrades to new major versions that may contain breaking
│ │ changes, we recommend adding version constraints in a required_providers block
│ │ in your configuration, with the constraint strings suggested below.
│ │ 
│ │ * hashicorp/azurerm: version = "~> 2.36.0"
│ │ 
│ │ Terraform has been successfully initialized!
│ │ Terraform runner "master-node" process exited.
│ └ terraform init ... (0.68 seconds)
│ 
│ ┌ terraform plan ...
│ │ Refreshing Terraform state in-memory prior to plan...
│ │ The refreshed state will be used to calculate this plan, but will not be
│ │ persisted to local or remote state storage.
│ │ 
│ │ data.azurerm_resource_group.kube: Refreshing state...                                                                                  ↵
│ │ [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter31aug]
│ │ data.azurerm_subnet.kube: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter31a ↵
│ │ ug/providers/Microsoft.Network/virtualNetworks/destroymeafter31aug/subnets/destroymeafter31aug]
│ │ azurerm_public_ip.master[0]: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroymeafter ↵
│ │ 31aug/providers/Microsoft.Network/publicIPAddresses/destroymeafter31aug-master-0]
│ │ azurerm_managed_disk.kubernetes_data: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destr ↵
│ │ oymeafter31aug/providers/Microsoft.Compute/disks/destroymeafter31aug-kubernetes-data-0]
│ │ azurerm_network_interface.master: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destroyme ↵
│ │ after31aug/providers/Microsoft.Network/networkInterfaces/destroymeafter31aug-master-0]
│ │ azurerm_linux_virtual_machine.master: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e556/resourceGroups/destr ↵
│ │ oymeafter31aug/providers/Microsoft.Compute/virtualMachines/destroymeafter31aug-master-0]
│ │ azurerm_virtual_machine_data_disk_attachment.kubernetes_data: Refreshing state... [id=/subscriptions/6bee52a1-03e5-43d6-8e38-a37d54c1e ↵
│ │ 556/resourceGroups/destroymeafter31aug/providers/Microsoft.Compute/virtualMachines/destroymeafter31aug-master-0/dataDisks/destroymeaft ↵
│ │ er31aug-kubernetes-data-0]
│ │ 
│ │ ------------------------------------------------------------------------
│ │ 
│ │ No changes. Infrastructure is up-to-date.
│ │ 
│ │ This means that Terraform did not detect any differences between your
│ │ configuration and real physical resources that exist. As a result, no
│ │ actions need to be performed.
│ │ Terraform runner "master-node" process exited.
│ └ terraform plan ... (12.54 seconds)
└ 🌱 ~ Terraform: Check state master-node for destroymeafter31aug-master-0 (13.23 seconds)
cluster:
  status: ok
node_templates:
- name: master
  status: ok
nodes:
- group: master
  name: destroymeafter31aug-master-0
  status: ok
[deckhouse] root@e1267ddd825a / # cat /deckhouse/candi/cloud-providers/azure/terraform-modules/base-infrastructure/main.tf | grep endpoints
  service_endpoints    = local.service_endpoints
```

</details>

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: cloud-provider-azure
type: feature
summary: Add `serviceEndpoints` parameter in a provider cluster configuration.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/azure'...

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/azure'...

File: candi/cloud-providers/azure/openapi/cluster_configuration.yaml [hunk 1]
```diff
@@ -103,6 +103,14 @@ apiVersions:
               description: |
                 The name of the VNet.
               type: string
+      serviceEndpoints:
+        description: |
+          The list of Service endpoints to associate with the subnet.
+          Virtual Network (VNet) service endpoint provides secure and direct connectivity to Azure services over an optimized route over the Azure backbone network. Endpoints allow you to secure your critical Azure service resources to only your virtual networks. Service Endpoints enables private IP addresses in the VNet to reach the endpoint of an Azure service without needing a public IP address on the VNet. More information about Service Endpoints can be found in the [official documentation](https://learn.microsoft.com/en-us/azure/virtual-network/virtual-network-service-endpoints-overview).
```
Comment (z9r5) [inline candi/cloud-providers/azure/openapi/cluster_configuration.yaml:109]:
```suggestion
          The list of Service endpoints to associate with the subnet.

          Virtual Network (VNet) service endpoint provides secure and direct connectivity to Azure services over an optimized route over the Azure backbone network. Endpoints allow you to secure your critical Azure service resources to only your virtual networks. Service Endpoints enables private IP addresses in the VNet to reach the endpoint of an Azure service without needing a public IP address on the VNet. More information about Service Endpoints can be found in the [official documentation](https://learn.microsoft.com/en-us/azure/virtual-network/virtual-network-service-endpoints-overview).
```

File: candi/cloud-providers/azure/openapi/doc-ru-cluster_configuration.yaml [hunk 2]
```diff
@@ -55,6 +55,10 @@ apiVersions:
             vnetName:
               description: |
                 Имя VNet.
+      serviceEndpoints:
+        description: |
+          Список конечных точек служб ассоциированных с подсетью.
+          Конечная точка службы для виртуальной сети обеспечивает возможность безопасного и прямого подключения к службам Azure по оптимизированному маршруту через магистральную сеть Azure. Конечные точки позволяют защищать критически важные ресурсы служб Azure в пределах отдельных виртуальных сетей. Через конечные точки служб частные IP-адреса в виртуальной сети могут достигать конечную точку службы Azure без необходимости использовать общедоступный IP-адрес в виртуальной сети. Больше информации можно найти в [официальной документации](https://learn.microsoft.com/ru-ru/azure/virtual-network/virtual-network-service-endpoints-overview).
```
Comment (z9r5) [inline candi/cloud-providers/azure/openapi/doc-ru-cluster_configuration.yaml:61]:
```suggestion
          Список конечных точек служб ассоциированных с подсетью.

          Конечная точка службы для виртуальной сети обеспечивает возможность безопасного и прямого подключения к службам Azure по оптимизированному маршруту через магистральную сеть Azure. Конечные точки позволяют защищать критически важные ресурсы служб Azure в пределах отдельных виртуальных сетей. Через конечные точки служб частные IP-адреса в виртуальной сети могут достигать конечную точку службы Azure без необходимости использовать общедоступный IP-адрес в виртуальной сети. Больше информации можно найти в [официальной документации](https://learn.microsoft.com/ru-ru/azure/virtual-network/virtual-network-service-endpoints-overview).
```

File: candi/cloud-providers/azure/openapi/doc-ru-cluster_configuration.yaml [hunk 3]
```diff
@@ -55,6 +55,11 @@ apiVersions:
             vnetName:
               description: |
                 Имя VNet.
+      serviceEndpoints:
+        description: |
+          Список конечных точек служб ассоциированных с подсетью.
+
+          Конечная точка службы для виртуальной сети обеспечивает возможность безопасного и прямого подключения к службам Azure по оптимизированному маршруту через магистральную сеть Azure. Конечные точки позволяют защищать критически важные ресурсы служб Azure в пределах отдельных виртуальных сетей. Через конечные точки служб частные IP-адреса в виртуальной сети могут достигать конечную точку службы Azure без необходимости использовать общедоступный IP-адрес в виртуальной сети. Больше информации можно найти в [официальной документации](https://learn.microsoft.com/ru-ru/azure/virtual-network/virtual-network-service-endpoints-overview).
```
Comment (kirkonru) [inline candi/cloud-providers/azure/openapi/doc-ru-cluster_configuration.yaml:62]:
```suggestion
          Список Service Endpoints, ассоциированных с подсетью.

          Service Endpoints для виртуальной сети позволяют прямо и безопасно подключаться к сервисам Azure по оптимизированному маршруту через магистральную сеть Azure. Они позволяют ограничить доступ к критически важным ресурсам сервисов Azure пределами отдельных виртуальных сетей. Через Service Endpoints приватные IP-адреса в виртуальной сети могут подключаться к эндпоинтам сервисов Azure, не используя публичный IP-адрес в виртуальной сети. Больше информации можно найти в [официальной документации](https://learn.microsoft.com/ru-ru/azure/virtual-network/virtual-network-service-endpoints-overview).
```

**Review by kirkonru — APPROVED** [review]

**Pull request merged**